frontend/controller/actions/identity.js

```diff
@@ -9,22 +9,173 @@ import {
+    } else {
+      // Map the decrypted keys to the required format
+      const secretKeys = decryptedKeys.map(key => ({ key: deserializeKey(key), transient: true }))
+
+      // Store the secret keys using the sbp function
+      await sbp('chelonia/storeSecretKeys', new Secret(secretKeys))
+    }
+  } catch (error) {
+    console.error('[processOldIekList] Error fetching or processing old keys:', error)
+  }
+}
+
+/**
+ * Appends a new IEK to the existing list of old IEKs.
+ *
+ * @param contractID - The ID of the contract.
+ * @param IEK - The encryption key object.
+ * @param oldIEK - The old IEK to be appended.
+ * @param encryptedData - The encrypted data string, or null if not available.
+ * @returns The updated encrypted data containing the new IEK.
+ * @throws {Error} - Throws an error if decryption of old IEK list fails.
+ */
+const appendToIekList = (contractID: string, IEK: Object, oldIEK: Object, encryptedData: ?string) => {
+  // Decrypt the old IEK list
+  const oldKeys = decryptOldIekList(contractID, oldIEK, encryptedData)
+
+  // If decryption fails, throw an error to prevent data loss
+  if (!oldKeys) {
+    throw new Error('Error decrypting old IEK list')
+  }
+
+  // Create a Set to store unique keys
+  const keysSet = new Set(oldKeys)
+
+  // Serialize the old IEK and add it to the Set
+  const serializedOldIEK = serializeKey(oldIEK, true)
+  keysSet.add(serializedOldIEK)
+
+  // Encrypt the updated list of keys and return the new encrypted data
+  const updatedKeysData = encryptedOutgoingDataWithRawKey(
+    IEK,
+    // Convert Set back to Array for serialization
+    JSON.stringify(Array.from(keysSet))
+  ).toString(`meta.private.oldKeys;${contractID}`)
+
+  return updatedKeysData // Return the updated encrypted data
+}
+
+// Event handler to detect password updates
+sbp('okTurtles.events/on', EVENT_HANDLED, (contractID, message) => {
+  const identityContractID = sbp('state/vuex/state').loggedIn?.identityContractID
+  // If the message isn't for our identity contract or it's not `OP_KEY_UPDATE`
+  // (possibly within `OP_ATOMIC`), we return early
+  if (contractID !== identityContractID || ![GIMessage.OP_ATOMIC, GIMessage.OP_KEY_UPDATE].includes(message.opType())) return
+
+  // If this could have changed our CSK, let's try to get the key ID and see if
+  // we have the corresponding secret key
+  const hasNewCsk = sbp('chelonia/contract/currentKeyIdByName', identityContractID, 'csk', true)
+  // If we do, we can still use the contract as normal
+  if (hasNewCsk) return
+
+  // Otherwise, force a logout
+  console.warn('Likely password change for identity contract. Logging us out.', identityContractID)
+  sbp('gi.actions/identity/logout').catch(e => {
+    console.error('Error while automatically logging out', e)
+  })
+})
+
 export default (sbp('sbp/selectors/register', {
   'gi.actions/identity/create': async function ({
-    IPK,
-    IEK,
+    IPK: wIPK,
```

Comment:
These changes aren't strictly related to this PR, rather something that slipped through the cracks in the past. This wraps secrets in `Secret` objects.